### ISSUE TITLE
[question] is it possible to stop parsing on the first error?

### ISSUE THREAD

**Issue by gerardolima** [colinhacks/zod#1403]
I'm using Zod to validate HTTP payloads, some of which may be expensive to validate. Is it possible for `schema.parse()` and their friends to stop on the first error and return a single ZodIssue?

**Comment by colinhacks** [comment]
Zod is designed to surface as many errors as possible, so not really. Some sort of `abortEarly` mode is certainly possible but I'm not convinced that it's worth the increase in complexity - it's not a very common use case.

**Comment by gerardolima** [comment]
hey, @colinhacks, first of all, **thank you for the great work** 💪🏽! I understand that's not the intended _original_ use case for Zod. I'm moving from [Joi](https://joi.dev/) -- which I like, but lacks the type inference Zod does so well -- and my idea was to migrate all schemas. My problem is that I've been also using Joi to validate HTTP payloads on a [Hapi](https://hapi.dev/tutorials/validation/) service and some of those validations can be expensive.

Maybe my case is too specific, and it wouldn't worth the changes, but maybe HTTP payload validation could be a new use case that Zod might also address :) Otherwise, and obviously, please, close the issue -- I don't want to bother you with things out of the scope of the project.

**Comment by jayarjo** [comment]
+1

Bailing out on very first error, should be as simple as throwing an exception the moment you stumble on one. No?

**Comment by srieas** [comment]
+1

**Comment by irrelevation** [comment]
There is support for [aborting early ](https://zod.dev/?id=abort-early) when using [.superRefine](https://zod.dev/?id=superrefine). It is more work but might fit your use case.

**Comment by gerardolima** [comment]
hey, @irrelevation thank you for pointing that out, but I think [.superRefine](https://zod.dev/?id=superrefine) avoids the main benefit of using Zod (or other type inference library), which is expressing types in a declarative way.

**Comment by carlBgood** [comment]
> Zod is designed to surface as many errors as possible, so not really. Some sort of `abortEarly` mode is certainly possible but I'm not convinced that it's worth the increase in complexity - it's not a very common use case.

I agree that the design is structured to expose as many errors as possible - however, I disagree that it's not a common use case to need to halt mid-validation.

A pretty common use-case is form-validation utilizing database queries. Example: if I want to validate a simple email/password form, I can `string().email()`. But if want to combine that with the database query to keep all my errors in sync and make it `string().email().superRefine()` I can only halt/abort after the first db query I run in the refinement - and if any previous validators in the chain failed - I'm running the db query for nothing - and either with potentially unsecure data or I need to _re-validate_ in the refinement to ensure it's good to go. The other alternative would be to use a refinement on the entire data object, but the same issue would persist - I'd be running all the validators and the first db query even if the entry data was invalid.

Just my two cents, but it would be a major improvement in an already amazing package. I can see huge value in a simple `bail()` function - a la https://github.com/colinhacks/zod/issues/1606#issuecomment-1381702201 - that would halt and return (akin to the `fatal` & `Z.NEVER` in refinements).

Thanks again, @colinhacks, for Zod - it's a great resource!

**Comment by JaneJeon** [comment]
Definitely still relevant and would be greatly appreciated

**Comment by scotttrinh** [comment]
I believe this is a case of dualing use-cases: people fought pretty hard for the current API of "surface as many errors as possible" since we used to fail fast.

> A pretty common use-case is form-validation utilizing database queries. Example: if I want to validate a simple email/password form, I can string().email(). But if want to combine that with the database query to keep all my errors in sync and make it string().email().superRefine() I can only halt/abort after the first db query I run in the refinement - and if any previous validators in the chain failed - I'm running the db query for nothing - and either with potentially unsecure data or I need to re-validate in the refinement to ensure it's good to go.

I think this very neatly describes why I avoid using the blessed internal validators: they're just too "special case"-y and require to much refactoring if you stray outside of their narrow use-case. If you need to control the parallelism of your validation logic, I would building up some composable `superRefine`-style functions and sticking with those.

Another option is to use `pipe` to abort earlier on `email` and put your more expensive refinements in the schema you're piping to. So:

```ts
const schema = z.object({
  foo: z
    .string()
    .email()
    .pipe(
      z.string().superRefine(() => {
        // super expensive calc
      })
    ),
});
```

That won't work for all use-cases and now you have to deal with the `ZodPipeline` type instead of just `ZodEffects`, but it might work for some large percentage of your usage.

**Comment by valentsea** [comment]
Plus. It would be a significant improvement for zod.

**Comment by TamirCode** [comment]
+1 
I only display the first error of each field

**Comment by bulicmatko** [comment]
In form validation, it's more common to display only one error for a field and not to throw multiple errors at our users.

For example:
- We have `string().min(1, "required").email()` schema for a field.
- User submits a form with an empty string.
- In most cases, we show only the first error to the user, which is `"required"` in this case.
- Showing "required" and "invalid email" is bad.

There's no point in validating if an empty string is an email.

When we add a few more "expensive" rules in the validation chain, for example, we ping a server, the issue becomes more apparent.

I can understand that `zod` is (maybe) not intended to cover these cases but as @valentsea mentioned above, it would be a significant improvement for devs that use it in form validation.

**Comment by TamirCode** [comment]
@gerardolima  Y u close 🫤

**Comment by JaneJeon** [comment]
Was this added in zod, or just closed because the original poster no longer needs it anymore?

**Comment by carlBgood** [comment]
Closed?!?

**Comment by gerardolima** [comment]
hey, @TamirCode, @JaneJeon and @carlBgood, I had closed this issue because I believe this use case is not intended to be tackled by Zod (which I think is fair). I reopened because your messages indicate this may be an important subject to you. I opened this ticket some time ago, and I've already solved my problem, so I won't really be an active part of this conversation. Cheers

**Comment by OctavioSerpe** [comment]
I have stumbled into the same issue, I need to bail validation on specific errors (like express-validator allows to https://express-validator.github.io/docs/api/validation-chain/#bail). One approach could be to do chain validations separately (tedious) or maybe just get the first error and do the heavy calculation on another schema. 

I believe it would be a nice addition to this incredible tool.

Cheers

**Comment by rafaell-lycan** [comment]
@colinhacks what would be the problem supporting something such as `abortEarly` like Joi/Yup does?

Furthermore, what would be the downside of it? What about performance improvements?

```ts
schema.parse(value, { abortEarly: true })
schema.safeParse(value, { abortEarly: true })

await schema.parseAsync(value, { abortEarly: true })
await schema.safeParseAsync(value, { abortEarly: true })
```

**Comment by gustawdaniel** [comment]
It is implemnted in `valibot` https://github.com/fabian-hiller/valibot/issues/18

commit: https://github.com/fabian-hiller/valibot/commit/37ac397666e2c8bc2529d98f50d861026fbacc93

interface:

```
abortPipeEarly?: boolean
```

**Comment by KikoCosmetics** [comment]
+1 guys. In my case, I have a refine that is an asyncronous call (to verify email addresses). 
It fires so many times even if an invalid email is typed. 
Great waste to my mind.
I'm trying using a superrefiner, but doesn't look like I can now that the field is already invalid...
Should I move all the validations there? Let's hope not!

**Comment by jyotirmaybarman** [comment]
+1 from my side

**Comment by danteissaias** [comment]
> In form validation, it's more common to display only one error for a field and not to throw multiple errors at our users.
> 
> For example:
> 
> * We have `string().min(1, "required").email()` schema for a field.
> * User submits a form with an empty string.
> * In most cases, we show only the first error to the user, which is `"required"` in this case.
> * Showing `"required"` and `"invalid email"` is kind of redundant.
> 
> Also, in this kind of form validation, there's no point in validating if an empty string is an email.
> 
> When we add a few more "expensive" rules in the validation chain, for example, we ping a server, the issue becomes more apparent.
> 
> I can understand that `zod` is (maybe) not intended to cover these cases but as @valentsea mentioned above, it would be a significant improvement for devs that use it in form validation.

Did you find a solution for this case?

**Comment by catinrage** [comment]
+1 it would be a great option to have, it would avoid duplicate validations.

**Comment by buzzy** [comment]
"it's not a very common use case." I really hope that was a joke? Doing database lookups as a part of validation is **very** common and should obviously not be done if previous validation rules failed. Luckily we now have this in valibot.

**Comment by MentalGear** [comment]
We were trying to use zod for general validation (client/server) with superforms but the inability to bail early on the first error prevents this, as valuable server resources are wasted when zod still tries to validate the rest of the data even though the first entry is already faulty.

Devs overwhelmingly want to have the option to activate this mode, I'm really wondering why the maintainers are so set against it. It's just an option, and it's always good to have options, right @colinhacks.

**Comment by gregmsanderson** [comment]
@MentalGear I've just been finding exactly the same issue as you. Very frustrating. 

I notice that another library superforms supports is VineJS which does appear to have a "bail" mode https://vinejs.dev/docs/schema_101#bail-mode

However _that_ currently throws a "node:dns" warning. Nothing is ever simple!

**Comment by jeslenbucci** [comment]
Like many here, I'm also queyring my database to check for email addresses in my validation and noticed that it was executing every time, even when the email was not valid. To get around this, I just added a bit of a mix from what the others mentioned.

Rather than chaining all the default checks before the superRefine(), I just do within it. Then use the fatal flag when using ctx.addIssue().

```
const emailSchema = z.string().superRefine(async (email, ctx) => {
  const emailTest = z.string().email().safeParse(email)

    if (!emailTest.success) {
      ctx.addIssue({
        code: ZodIssueCode.invalid_string,
        validation: 'email',
        message: 'Invalid Email',
        fatal: true,
      })
      return z.NEVER
    }
    
    // Do email query here
})
```

**Comment by jeslenbucci** [comment]
Here's an even more complex example that I have working with react-hook-forms with a dynamic form builder I created. What I do is first create the basic schema as a constant. Then, in the super refinement, I do a safe parse. If it's not successful, I loop over each issue and spread the error into th ectx.addIssue() method. Then I return z.Never. This recreates the same basic validation that you'd get without the refinement.

In my example, I'm parsing a large address object that allows for both mailing and physical addresses. I add an additional `uuid` prop to my object because that's the way it's set up in my forms. Otherwise, you wouldn't need to shape your object as I do with my `uuidSchema`

```javascript
  // define the basic schema
  const basicAddressSchema = z.object({
    mailing: z.object({
      line1: z.string().min(2),
      line2: z.string().optional(),
      city: z.string().min(2),
      subdivision: z.string().min(2),
      postalCode: z.string().min(2),
      country: z.string().min(2),
    }),
    physical: z.object({
      isSameAsMailing: z.boolean(),
      line1: z.string().min(2),
      line2: z.string().optional(),
      city: z.string().min(2),
      subdivision: z.string().min(2),
      postalCode: z.string().min(2),
      country: z.string().min(2),
    }),
  })
  
  // set up actual address schmea with the super refine method
  const addressSchema = basicAddressSchema.superRefine((address, ctx) => {
    // create my uuid schema, you would need to match your actual data
    const uuidSchema = z.object({
      [uuid]: z.object({
        address: basicAddressSchema,
      }),
    })

    // parse the schema
    const uuidValidation = uuidSchema.safeParse({ [uuid]: { ...address } })

    // if there are basic errosrs
    if (!uuidValidation.success) {
      // iterate over each
      uuidValidation.error.issues.forEach((error) => {
        // add them to the ctx
        ctx.addIssue({
          ...error,
        })
      })
      //
      return z.NEVER
    }
    
    console.log("this will not appear if there are errors")
  })
```

**Comment by rhaeyx** [comment]
> In form validation, it's more common to display only one error for a field and not to throw multiple errors at our users.
> 
> For example:
> 
>     * We have `string().min(1, "required").email()` schema for a field.
> 
>     * User submits a form with an empty string.
> 
>     * In most cases, we show only the first error to the user, which is `"required"` in this case.
> 
>     * Showing `"required"` and `"invalid email"` is kind of redundant.
> 
> 
> Also, in this kind of form validation, there's no point in validating if an empty string is an email.
> 
> When we add a few more "expensive" rules in the validation chain, for example, we ping a server, the issue becomes more apparent.
> 
> I can understand that `zod` is (maybe) not intended to cover these cases but as @valentsea mentioned above, it would be a significant improvement for devs that use it in form validation.

+1 to this, I have the same use case with this reply. Has this been added or are there no plans in adding this?

**Comment by sebastien-comeau** [comment]
+1

**Comment by buzzy** [comment]
> Zod is designed to surface as many errors as possible, so not really. Some sort of `abortEarly` mode is certainly possible but I'm not convinced that it's worth the increase in complexity - it's not a very common use case.

Not a common use-case? That is one of the weirdest things I ever read! I would argue the opposite. Why would you WANT to keep validating a field that already failed validation? If a field is set to "required", why would you need to also check if it's a valid e-mail? Makes absolutely no sense. Abort early should be the DEFAULT. What exactly would you do with all those validations that fails for an empty string? Show 5 failed validation messages to the user? Crazy.

**Comment by dmgauthier** [comment]
+1

**Comment by buzzy** [comment]
+1

**Comment by abdurahmanshiine** [comment]
+1

**Comment by AlexanderBich** [comment]
+1

**Comment by yousufiqbal** [comment]
Sakht londa he. Kr de bhai implement.

**Comment by MaximusHaximus** [comment]
+1 -- I can't understand why there is so much resistance to adding this feature.

**Comment by mhdalmajid** [comment]
+1
I am surprised zod dose not have abort . 
we love zod anyway

**Comment by romanstetsyk** [comment]
The absence of early abort is basically the only reason I don't use Zod.

**Comment by rafaell-lycan** [comment]
+1

**Comment by MentalGear** [comment]
It think it should be clear by now that the maintainer won't receive updates on this closed issue any more.
If we want this feature, we need to open a new issue, reference this one, and ask for the next major version to support this feature.

**Comment by buzzy** [comment]
> It think it should be clear by now that the maintainer won't receive updates on this closed issue any more. If we want this feature, we need to open a new issue, reference this one, and ask for the next major version to support this feature.

This ticket looks very much open to me...

**Comment by mhdalmajid** [comment]
> It think it should be clear by now that the maintainer won't receive updates on this closed issue any more.
> 
> If we want this feature, we need to open a new issue, reference this one, and ask for the next major version to support this feature. 

🤔, it still open yet

**Comment by catinrage** [comment]
> Abort early should be the DEFAULT.
I can't agree more, imagine you have a list of expensive validations, but first 2 or 3 are just some format validations, it would be great to ignore there and avoid those expensive ones.

**Comment by buzzy** [comment]
> > In form validation, it's more common to display only one error for a field and not to throw multiple errors at our users.
> > For example:
> > 
> > * We have `string().min(1, "required").email()` schema for a field.
> > * User submits a form with an empty string.
> > * In most cases, we show only the first error to the user, which is `"required"` in this case.
> > * Showing `"required"` and `"invalid email"` is kind of redundant.
> > 
> > Also, in this kind of form validation, there's no point in validating if an empty string is an email.
> > When we add a few more "expensive" rules in the validation chain, for example, we ping a server, the issue becomes more apparent.
> > I can understand that `zod` is (maybe) not intended to cover these cases but as @valentsea mentioned above, it would be a significant improvement for devs that use it in form validation.
> 
> This is the exact issue I am having right now. I have over 15 forms with their UI already built, trying to do frontend/backend validation and integration only to discover this roadblock and I'd have to resort to using verbose and less readable superRefine even for build-in validation methods.

Yes. It is extremely strange that this validation library is as popular as it is with this obvious limitation.

**Comment by MentalGear** [comment]
I assume the maintainer has unsubscribed a long time ago from this topic. We need to add a new one, summarize this topic, and reference this one, before next major version is set/released soon.

**Comment by mhdalmajid** [comment]
new Issue has been created https://github.com/colinhacks/zod/issues/3884

**Comment by sbreese** [comment]
Here is my solution to this problem:

`    let flattenedZodErrors = validatedFields.error.flatten().fieldErrors;

    /* flattenedZodErrors BEFORE: {
      { zip: ['Zip code is required.', 'Zip code is invalid.'] }
    */
    if ("zip" in flattenedZodErrors) {
      flattenedZodErrors = {
        ...flattenedZodErrors,
        zip: [flattenedZodErrors.zip![0]],
      };
    }
    /* flattenedZodErrors AFTER: {
      { zip: ['Zip code is required.'] }
    */`

**Comment by buzzy** [comment]
> My solution is to clean up the errors object after Zod validation.
> 
> ```
> let flattenedZodErrors = validatedFields.error.flatten().fieldErrors;
> /* flattenedZodErrors BEFORE:
>   { zip: ['Zip code is required.', 'Zip code is invalid.'] }
> */
> if ("zip" in flattenedZodErrors) {
>   flattenedZodErrors = {
>     ...flattenedZodErrors,
>     zip: [flattenedZodErrors.zip![0]],
>   };
> }
> /* flattenedZodErrors AFTER: {
>   { zip: ['Zip code is required.'] }
> */`
> ```

But that has never been the issue? The issue is that all the "expensive" validations are run, even though they don't have to. If the "required" validation fails, it should not need to run all the other validation steps that might include database lookups etc.

**Comment by ruanswanepoel** [comment]
Using Zod to parse millions of records of CSV data across hundreds of files. I don't need to know all of the errors, I just need to know whether it parsed properly or not.

Abort early would be very much appreciated.

**Comment by MentalGear** [comment]
Still very relevant.

**Comment by dewita97** [comment]
Hi guys,

I found something interesting in the library repository and in the documentation. 

Since May 2025, the abortEarly option has been commented out in the ts schema: [link](https://github.com/colinhacks/zod/blob/main/packages/zod/src/v4/core/schemas.ts)

Here is the related commit: [link](https://github.com/colinhacks/zod/commit/85928549a7c2aff851d8625588ed7aa2e394831f#diff-f84405ad925758ccb224a20fe7ca8c6ee375d7faa88431059e008077550f8cd9)

If abortEarly is present in the file, chances are we'll get that option one day!

**Comment by em1382** [comment]
This would be a huge plus for those using Zod on the server-side!

**Comment by colinhacks** [comment]
Superseded by #3884, which is where this continues. My read hasn't changed: zod surfaces as many errors as possible by design, and a parse-level `abortEarly` is more complexity than the use case carries.
